python/ray/_private/runtime_env/utils.py

```diff
@@ -280,3 +286,89 @@ def from_dict(
         _build_proto_container_runtime_env(runtime_env_dict, proto_runtime_env)
         _build_proto_plugin_runtime_env(runtime_env_dict, proto_runtime_env)
         return cls(proto_runtime_env=proto_runtime_env)
+
+
+class SubprocessCalledProcessError(subprocess.CalledProcessError):
+    """The subprocess.CalledProcessError with stripped stdout."""
+
+    LAST_N_LINES = 10
+
+    @staticmethod
+    def _get_last_n_line(str_data: str, last_n_lines: int) -> str:
+        if last_n_lines < 0:
+            return str_data
+        lines = str_data.split("\n")
+        return "\n".join(lines[-last_n_lines:])
+
+    def __str__(self):
+        str_list = [super().__str__().strip(",.")]
+        out = {
+            "stdout": self.stdout,
+            "stderr": self.stderr,
```

Comment:
"stderr" will be empty because we redirect it to stdout https://github.com/ray-project/ray/pull/22381/files#diff-3968b5fc7121d304a253988b1bed9fdd58c910dbae0eb33bc0405e2a9a1bfec3R346 ?

Reply:
> this part

Yes. But we can't assume that the stderr is always redirected to the stdout. The implementation of `SubprocessCalledProcessError` should cover the interface of the parent `subprocess.CalledProcessError`.

Reply:
> This looks very similar to [this part](https://github.com/ray-project/ray/blob/master/python/ray/_private/runtime_env/conda_utils.py#L216), maybe you can modify the implementation in conda_utils.py to avoid splitting the implementation

The `exec_cmd` and `exec_cmd_stream_to_logger` in the `conda_utils.py` are used by many runtime envs, not only by the conda. I thinks it's better to put the exec utils into the `utils.py` instead of `conda_utils.py`.

But, we are going to make all the runtime env async, so `exec_cmd` and `exec_cmd_stream_to_logger` will be removed after they are not referenced.